# CFIA-NCFAD/scovtree: Output

## Introduction

This document describes the output produced by the pipeline.

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level results directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/)
and processes data using the following steps:

1. [Prepare Input Sequences](#prepare-input-sequences)
2. [Pangolin](#pangolin) lineage assignment
3. [Filtering GISAID sequences](#filtering-gisaid-sequences) (if specified)
4. [Multiple sequence alignment](#multiple-sequence-alignment)
5. [Nextclade](#nextclade) Mutation Calling
6. [Phylogenetic tree inference](#phylogenetic-tree-inference)
7. [Phylogenetic tree with SNPs](#phylogenetic-tree-with-SNPs)
8. [shiptv tree visualization](#shiptv-tree-visualization)

### Prepare Input Sequences

Input sequence IDs may not be preserved by [Pangolin]. The text after the first pipe (`|`) or whitespace character is trimmed away so that the results can be linked to the original sequences and tree taxa.

<details markdown="1">
<summary>Output files</summary>

* `input_sequences/input_sequences.correctedID.fasta`:

</details>

### Pangolin

Each input sequence is assigned to a SARS-CoV-2 lineage with [Pangolin]. This lineage info is used for filtering GISAID sequences and also shown in the [shiptv] tree.

<details markdown="1">
<summary>Output files</summary>

* `pangolin/pangolin.csv`: [Pangolin] lineage assignment results.

</details>

### Filtering GISAID sequences

If both [GISAID] sequences (`--gisaid_sequences`) and metadata (`--gisaid_metadata`) are specified, the [GISAID] sequences belonging to the same [Pangolin] lineages as the `--input` sequences will be filtered for. Poor quality sequences will be filtered out (too long; too short; too many ambiguous sites).

<details markdown="1">
<summary>Output files</summary>

* `gisaid/`
  * `gisaid_sequences.filtered.fasta`: [GISAID] sequences filtered based on metadata, quality and [Pangolin] lineage.
  * `gisaid_metadata.filtered.tsv`: Metadata for filtered [GISAID] sequences.
  * `gisaid_metadata.nextstrain.tsv`: Metadata for filtered [GISAID] sequences compatible with [Nextstrain] analysis.
  * `gisaid_filtering_stats.json`: GISAID filtering stats JSON.

</details>

### Multiple sequence alignment

Multiple sequence alignment (MSA) of user specified sequences and GISAID sequences (if specified) is performed using either [MAFFT] or [Nextalign].

<details markdown="1">
<summary>Output files</summary>

* `msa/`
  * `msa.filtered.fasta`: Filtered MSA sequences for phylogenetic analysis.
  * `metadata.filtered.tsv`: Metadata for filtered MSA sequences.
  * `nextalign/`:
    * `sequences.nextalign.fasta`: Nextalign MSA output FASTA file.
    * `nextalign.insertions.csv`: Nextalign insertions present in input sequences relative to reference sequence.

</details>

### Nextclade

[Nextclade] analysis is performed to identify amino acid (AA) mutations.

<details markdown="1">
<summary>Output files</summary>

* `nextclade/`
  * `aa_mutation_matrix.tsv`: An aa mutation matrix of sequences, `1` if aa mutation represented in sequences, otherwise is `0`
  * `nextclade.csv`: The `csv` file is generated by [Nextclade]
  * `sequences.nextclade.fasta`: Sequences for running [Nextclade]
  
</details>

### Phylogenetic tree inference

Maximum-likelihood phylogenetic tree is generated from the MSA of your input sequences and related GISAID sequences (if specified) using [IQ-TREE].

<details markdown="1">
<summary>Output files</summary>

* `iqtree/`
  * `iqtree-*.treefile`: Newick format IQ-TREE phylogenetic tree
  * `iqtree-*.iqtree`: IQ-TREE phylogenetic analysis report
  * `iqtree-*.log`: IQ-TREE log file
  * `iqtree-*.mldist`: IQ-TREE maximum-likelihood distances output file
  
</details>

### Phylogenetic tree with SNPs

An R [ggtree] phylogenetic tree with SNPs highlighted can be generated if no GISAID sequence and metadata is specified.

<details markdown="1">
<summary>Output files</summary>

* `plots/`
  * `phylogentic_tree_snps.pdf`: A phylogenetic tree with SNPs highlighted.
  * `data/`:
    * `alleles.tsv`: A `tsv` file with variants found in samples.
  
</details>
  
## shiptv tree visualization

An interactive phylogenetic tree visualization in a standalone HTML is produced with [shiptv].

<details markdown="1">
<summary>Output files</summary>

* `shiptv/`
  * `leaflist`: The subset of taxa visualized in the shiptv tree.
  * `metadata.leaflist.tsv`: Taxa metadata table for selected taxa.
  * `metadata.merged.tsv`: `metadata.leaflist.tsv` is merged with Pangolin lineage report and, optionally, the Nextclade amino acid mutation matrix.
  * `metadata.shiptv.tsv`: The metadata file is generated by [shiptv].
  * `shiptv.html`: shiptv interactive phylogenetic tree visualization as a standalone HTML file.

</details>

## Pipeline information

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.

<details markdown="1">
<summary>Output files</summary>

* `pipeline_info/`
  * Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  * Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.tsv`.

</details>

[ggtree]: https://bioconductor.org/packages/release/bioc/html/ggtree.html
[GISAID]: https://www.gisaid.org/
[IQ-TREE]: http://www.iqtree.org/
[jts/ncov-tools]: https://github.com/jts/ncov-tools
[MAFFT]: https://mafft.cbrc.jp/alignment/software/
[Nextalign]: https://github.com/nextstrain/nextclade/tree/master/packages/nextalign_cli
[Nextclade]: https://github.com/nextstrain/nextclade/tree/master/packages/nextclade_cli
[Nextflow]: https://www.nextflow.io/
[Nextstrain]: https://nextstrain.org/
[nf-core]: https://nf-co.re/
[Pangolin]: https://github.com/cov-lineages/pangolin/
[SARS-CoV-2]: https://www.ncbi.nlm.nih.gov/nuccore/MN908947.3/
[shiptv]: https://github.com/peterk87/shiptv
